### ISSUE TITLE
`action:subaction` and related actions' times

### ISSUE THREAD

**Issue by ajnelson-nist** [ucoProject/UCO#557]
Let's say we have the following graph, depicting an action that starts at 3AM (UTC) and ends at 4AM UTC:

```json
{       
    "@context": {
        "kb": "http://example.org/kb/",
        "action": "https://ontology.unifiedcyberontology.org/uco/action/",
        "xsd": "http://www.w3.org/2001/XMLSchema#"
    },          
    "@graph": [     
        {
            "@id": "kb:Action-1",
            "@type": "action:Action",
            "uco-action:subaction": {
                "@id": "kb:Action-2",
                "@type": "action:Action"
            },
            "uco-action:startTime": {
                "@type": "xsd:dateTime",
                "@value": "2020-01-02T03:00:00Z"
            },
            "uco-action:endTime": {
                "@type": "xsd:dateTime",
                "@value": "2020-01-02T04:00:00Z"
            }
        }
    ]
}
```

Do we know the following about `kb:Action-2`:

1. Its start time is at or after `2020-01-02T03:00:00Z`.
2. Its end time is before or at `2020-01-02T04:00:00Z`.

Generally speaking, would a subaction always be time-bounded?

**Comment by plbt5** [comment]
As far as I can determine from the Action ontology, there are no constraints that limit a subaction to start before or end after its parent action. This is in conflict with the definition for the subaction: "[a subaction] References to other actions that make up part of a larger more complex action." that suggests that the parent (larger more complex) action cannot start after the first, nor end before the last, subaction.

In conclusion: whether or not it is the intention for the parts to be proper subparts of the whole, the intention of the textual definition of the parts do not match the logical construction of the whole by its parts. If anything, this conflict should be resolved one way or another.

**Comment by ajnelson-nist** [comment]
@plbt5 - Does it it look like `action:subaction` could be a specialization of `isPerdurantProperPartOf` from [Issue 544](https://github.com/ucoProject/UCO/issues/544)?

**Comment by plbt5** [comment]
Indeed it seems to look that way. We need to assure, though, that by introducing this we do *not* imply that `isPerdurantProperPartOf` is an `InverseFunctionalProperty`. With that I mean that while an `action:subaction` could be considered to be a sub-action of only one super-action, this is not the case with the proper-part relation. 

At least, that is what I conclude from the [supplemental principles](https://plato.stanford.edu/Archives/spr2005/entries/mereology/#3) in mereology.

**Comment by ajnelson-nist** [comment]
@plbt5 The concern on `owl: InverseFunctionalProperty` in subproperties is fair.  My understanding is that if a property is not defined as inverse-functional, but a subproperty is, then the parent property does not become inverse-functional.

I just did a test[^1] with [Protege](https://protege.stanford.edu/) and [Robot](https://robot.obolibrary.org/), which both provide access to some Java-based reasoners.  I confirmed my understanding with the following ontology.  Protege complains of inconsistency when a reasoner is turned on, appropriately, until the commented lines are adjusted like suggested.  The inconsistency is that `kb:Thing-2` and `kb:Thing-3` are inferred to be the same individual via an `owl:FunctionalProperty`, and that individual inherits both of the different `ex:name` values, becoming inconsistent with `ex:ThingWithOneName`'s restriction of just one name value.  `kb:Thing-5` and `kb:Thing-6` trigger the same kind of inconsistency via an `owl:InverseFunctionalProperty`.  The example works if the other subproperty, `ex:op2`, is used instead, because the subproperties `ex:op3` and `ex:op4` do not induce their shared superproperty `ex:op1` to become a functional or inverse-functional property.

ROBOT ([1.9.5](https://github.com/ontodev/robot/releases/tag/v1.9.5)) complains appropriately[^2] with either of the following commands:

```bash
java -jar robot.jar \
	  reason \
	  --input asserted.ttl \
	  --output reasoned-hermit.ttl \
	  --reasoner HermiT
java -jar robot.jar \
	  reason \
	  --input asserted.ttl \
	  --output reasoned-jfact.ttl \
	  --reasoner JFact
```

```turtle
@prefix ex: <http://example.org/ontology/> .
@prefix kb: <http://example.org/kb/> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

ex:ThingWithOneName
	a owl:Class ;
	rdfs:subClassOf [
		a owl:Restriction ;
		owl:onProperty ex:name ;
		owl:cardinality "1"^^xsd:nonNegativeInteger ;
	] ;
	.

ex:name
	a owl:DatatypeProperty ;
	rdfs:range xsd:string ;
	.

ex:op1
	a owl:ObjectProperty ;
	rdfs:domain ex:ThingWithOneName ;
	rdfs:range ex:ThingWithOneName ;
	.

ex:op2
	a owl:ObjectProperty ;
	rdfs:subPropertyOf ex:op1 ;
	.

ex:op3
	a
		owl:ObjectProperty ,
		owl:FunctionalProperty
		;
	rdfs:subPropertyOf ex:op1 ;
	.

ex:op4
	a
		owl:ObjectProperty ,
		owl:InverseFunctionalProperty
		;
	rdfs:subPropertyOf ex:op1 ;
	.

kb:Thing-1
	a owl:Thing ;
	.

kb:Thing-2
	a owl:Thing ;
	ex:name "2" ;
	.

kb:Thing-3
	a owl:Thing ;
	ex:name "3" ;
	.

kb:Thing-4
	a owl:Thing ;
	.

kb:Thing-5
	a owl:Thing ;
	ex:name "5" ;
	.

kb:Thing-6
	a owl:Thing ;
	ex:name "6" ;
	.

########################################################################
# Below this line, if the ex:op3 and ex:op4 predicates are replaced with
# ex:op2, the ontology is consistent.
########################################################################

kb:Thing-1
	ex:op3
		kb:Thing-2 ,
		kb:Thing-3
		;
	.

kb:Thing-5
	ex:op4 kb:Thing-4 ;
	.

kb:Thing-6
	ex:op4 kb:Thing-4 ;
	.

#kb:Thing-1
#	ex:op2
#		kb:Thing-2 ,
#		kb:Thing-3
#		;
#	.
#
#kb:Thing-5
#	ex:op2 kb:Thing-4 ;
#	.
#
#kb:Thing-6
#	ex:op2 kb:Thing-4 ;
#	.
```

Protege provides the following explanation dialogue:

![protege-fp-ifp-explanations](https://github.com/ucoProject/UCO/assets/11739701/0add09d2-3f35-4837-b20c-a5ad7e905c17)

[^1]: Participation by NIST in the creation of the documentation of mentioned software is not intended to imply a recommendation or endorsement by the National Institute of Standards and Technology, nor is it intended to imply that any specific software is necessarily the best available for the purpose.
[^2]: The default reasoner in ROBOT, ELK, does not raise notice of any inconsistencies.  It reviews according to the [OWL 2 EL Profile](https://www.w3.org/TR/owl2-profiles/#OWL_2_EL), which does not support `owl:FunctionalProperty` or `owl:InverseFunctionalProperty`, so this attempted use of ELK is out of scope of its functionality.

**Comment by ajnelson-nist** [comment]
I'd be comfortable proposing (separately) that `action:subaction` become an `owl:InverseFunctionalProperty`, and seeing if counter-examples come up in discussion.

Meanwhile, there is still the original question on this Issue, of subaction times relative to the superaction.  For now, I'm interested in a yes or no response on whether the superaction's and subaction's times behave in a manner analagous to, or exactly like, [`time:intervalIn`](https://www.w3.org/TR/owl-time/#time:intervalIn) (distinct from [`time:intervalDuring`](https://www.w3.org/TR/owl-time/#time:intervalDuring)).

**Comment by sbarnum** [comment]
Yes, I would assert that the intent of the design of Action and subaction is that the time bounds of any sub Action would fall within the time bounds of the super Action.

**Comment by ajnelson-nist** [comment]
Thank you, @sbarnum .

This Issue will remain open while awaiting an implementation that can encode that constraint.  At the present time, it is not possible to encode, because of some un-constrainable details on `action:endTime` and `action:startTime` as literals of datatype `xsd:dateTime`.  As `xsd:dateTime`s, they are not required to have a time zone, and if they have one, they are not required to have any particular time zone.  So the only comparison mechanism available currently, lexical (string) comparison, would only work when the time values are confirmed to have the same time zone.  This is not a test-functionality available in SHACL.  It is available in SPARQL with the [`TZ()`](https://www.w3.org/TR/sparql11-query/#func-tz) function, but it would take some runtime inferencing and SHACL-SPARQL testing to make work---and it would likely cease to work once user data reaches several common complexities (e.g., data from two time zones, data spelling UTC as `Z` vs `-00:00`).

There are other ways to test the superaction-subaction time bounding, but they would require reified time instants.  Any of the following could work, but each would be predicated on another not-yet-filed proposal:

```turtle
# Assuming adoption of OWL-Time: https://www.w3.org/TR/owl-time/
action:subaction rdfs:subPropertyOf time:intervalIn .
```

```turtle
# Assuming an outcome of Issue 544.
action:subaction rdfs:subPropertyOf core:isPerdurantProperPartOf .
```

([Issue 544](https://github.com/ucoProject/UCO/issues/544).)

**Comment by ajnelson-nist** [comment]
Testing for [Issue 651](https://github.com/ucoProject/UCO/issues/651) is exercising this alignment:

```turtle
uco-action:subaction rdfs:subPropertyOf uco-time:intervalIn .
```

A point of clarification on why `intervalIn` is being used instead of `intervalDuring`: `intervalDuring` sets a strict separation between the beginning and ending instants of the related proper intervals.  `intervalIn` is a parent property of `intervalDuring`, `intervalStarts`, and `intervalFinishes`.  (So, `kb:PI1 uco-time:intervalStarts kb:PI2` implies `kb:PI1 uco-time:intervalIn kb:PI2`.)  [OWL-Time Figure 2](https://www.w3.org/TR/owl-time/#fig-thirteen-elementary-possible-relations-between-time-periods-af-97) illustrates this.

The net effect: With the alignment made to `intervalIn`, it is possible for a subaction's start to coincide with the superaction's start, or likewise with the end.  However, `intervalEquals` is defined as property-disjoint with `intervalIn`, so a subaction's time can't wholly comprise the superaction's time.

**Comment by ajnelson-nist** [comment]
I also found during the alignment testing that the alignment may need to be a UCO 2.0.0 topic.

Reasoning: The Allen relation properties (`intervalIn`, `intervalStarts`, etc.) are all defined with domain and range as proper-intervalic perdurants - i.e., perdurants of >0 duration.  [Issue 651](https://github.com/ucoProject/UCO/issues/651) is aligning `uco-action:Action` under the top perdurant, but doesn't go further to say whether every action must be >0 duration.

I'm not sure we have explored whether there are instantaneous actions.  I am also working on a proposal for instantaneous events (drawing from PROV-O, coming after 651), but I don't currently have a firm stance one way or the other on instantaneous actions.

But, if we align subaction to intervalIn, then we'll have a slightly tricky validation rule case that any `Action` using `subaction` must be proper-intervalic---and I believe this is backward-incompatible with UCO 1.x.0.  It's still possible to write this without inducing an inference requirement, but it'll be a little cumbersome in spelling.
